src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1381 @@
+                {
+                    Logger.Append("  => ");
+                    Logger.Append(pattern.Syntax.ToString());
+                    Logger.Append(",");
+                    if (wasReported)
+                    {
+                        Logger.Append(" [redundant]");
+                    }
+                }
+
+                Logger.AppendLine();
+            }
+#endif
+        }
+
+        private static void CheckOrAndAndReachability(
+            ArrayBuilder<StateForCase> previousCases,
+            int patternIndex,
+            BoundPattern pattern,
+            DecisionDagBuilder builder,
+            BoundDagTemp rootIdentifier,
+            SyntaxNode syntax,
+            BindingDiagnosticBag diagnostics,
+            PooledHashSet<SyntaxNode> redundantNodes)
+        {
+            var context = new ReachabilityAnalysisContext(previousCases, patternIndex, builder, rootIdentifier, syntax, redundantNodes);
+
+            try
+            {
+                var normalizedPattern = PatternNormalizer.Rewrite(pattern, rootIdentifier.Type);
+
+#if ROSLYN_TEST_REDUNDANT_PATTERN
+                context.Logger.AppendLine($"Pattern: {pattern.Syntax.ToString()}");
+                context.Logger.AppendLine($"Normalized pattern: {normalizedPattern.DumpSource()}");
+#endif
+
+                analyze(normalizedPattern, context);
+
+                // The set of redundancies in pattern `A` is identical to the set of redundancies in pattern `not A`
+                // but our method for detecting redundancies only detects those in `or` cases (which we bring to the top after normalization).
+                // By analyzing `not A` too we can report more redundancies.
+                // For example: `if (o is not (<any pattern including a redundancy>))`, `if (i is 42 and not 43)` (which is the negation of `not 42 or 43`)
+                var negated = new BoundNegatedPattern(pattern.Syntax, negated: pattern, pattern.InputType, narrowedType: pattern.InputType);
+                var normalizedNegatedPattern = PatternNormalizer.Rewrite(negated, rootIdentifier.Type);
+
+#if ROSLYN_TEST_REDUNDANT_PATTERN
+                context.Logger.AppendLine($"Normalized negated pattern: {normalizedNegatedPattern.DumpSource()}");
+#endif
+
+                analyze(normalizedNegatedPattern, context);
+
+#if ROSLYN_TEST_REDUNDANT_PATTERN
+                string log = context.Logger.ToString();
+                // For debugging, uncomment the preprocessing directive and set a breakpoint below
+                ;
+#endif
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                diagnostics.Add(ErrorCode.HDN_RedundantPatternStackGuard, pattern.Syntax);
+            }
+
+            return;
+
+            static void populateStateForCases(ArrayBuilder<BoundPattern> set, PooledHashSet<LabelSymbol> labelsToIgnore,
+                ref TemporaryArray<StateForCase> casesBuilder, ReachabilityAnalysisContext context)
```

Comment:
>context [](http://example.com/codeflow?start=91&length=7)

Do we intend to mutate this value? Otherwise, consider using `in` modifier for this parameter.

Reply:
I believe `TemporaryArray` contains 4 fields of the element type, which are used to avoid allocating a builder, and `Add` can write to those fields. So I wouldn't expect `ref readonly TemporaryArray` to work in this context.